onnxruntime/core/providers/cuda/tensor/scatter_nd.h

```diff
@@ -3,18 +3,63 @@
 
 #pragma once
 
+#include <string>
 #include "core/providers/shared_library/provider_api.h"
 #include "core/providers/cuda/cuda_kernel.h"
+#include "core/providers/cuda/tensor/scatter_nd_kind.h"
 #include "core/providers/cpu/tensor/scatter_nd.h"
 
 namespace onnxruntime {
 namespace cuda {
 
-class ScatterND final : public CudaKernel {
+/**
+ * This implementation assumes there is common indices and
+ * reduction is not needed. The code does not check that condition.
+ * However in that case, the same output element could be accessed
+ * from different threads at the same time and the final value
+ * is unlikely to be correct.
+ */
+class ScatterNDDisjointAndNoReduction final : public CudaKernel {
  public:
-  explicit ScatterND(const OpKernelInfo& info) : CudaKernel(info) {}
+  explicit ScatterNDDisjointAndNoReduction(const OpKernelInfo& info) : CudaKernel(info) {}
   Status ComputeInternal(OpKernelContext* context) const override;
 };
 
+/**
+ * This is an implementation derived from the first one.
+ * It does atomic operation to handle conflicts.
+ * The result is unlikely to be correct if the reduction is none
+ * as there is no guarantee that the final value will be the one
+ * corresponding to the highest visited index.
+ * TODO: change the implementation of avoid conflicts.
+ */
+class ScatterNDWithAtomicReduction final : public CudaKernel {
+ public:
+  explicit ScatterNDWithAtomicReduction(const OpKernelInfo& info) : CudaKernel(info) {
+    std::string reduction = info.GetAttrOrDefault<std::string>("reduction", "none");
+
+    if (info.GetAttr<std::string>("reduction", &reduction).IsOK()) {
```

Comment:
Does this query the attribute the second time?

Reply:
I missed it. Fixed.